pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+	if err != nil {
+		return err
+	}
+	if credSecret == nil {
+		return migapi.CredSecretNotFound
+	}
+	secret.Data = map[string][]byte{
+		"access_key": []byte(credSecret.Data[migapi.AwsAccessKeyId]),
+		"secret_key": []byte(credSecret.Data[migapi.AwsSecretAccessKey]),
+	}
+	return nil
+}
+
+// Get an existing credentials Secret on the source cluster.
+func (t *Task) getRegistryCredentialsSecret(client k8sclient.Client) (*kapi.Secret, error) {
+	list := kapi.SecretList{}
+	labels := t.PlanResources.MigPlan.GetCorrelationLabels()
+	err := client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(labels),
+		&list)
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, nil
+}
+
+// Determine if two registry credentials secrets are equal.
+// Returns `true` when equal.
+func (t *Task) equalsRegistryCredentialsSecret(a, b *kapi.Secret) bool {
+	return reflect.DeepEqual(a.Data, b.Data)
+}
+
+// Ensure the imagestream for the migration registry on the specified cluster has been created
+func (t *Task) ensureRegistryImageStream(client k8sclient.Client, registryResources *MigrationRegistryResources) error {
+	registrySecret := registryResources.CredSecret
+	if registrySecret == nil {
+		return errors.New("migration registry credentials not found")
+	}
+	newImageStream, err := t.buildRegistryImageStream(registrySecret.GetName())
+	if err != nil {
+		return err
+	}
+	foundImageStream, err := t.getRegistryImageStream(client)
+	if err != nil {
+		return err
+	}
+	if foundImageStream == nil {
+		registryResources.ImageStream = newImageStream
+		err = client.Create(context.TODO(), newImageStream)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.ImageStream = foundImageStream
+	if !t.equalsRegistryImageStream(newImageStream, foundImageStream) {
+		t.updateRegistryImageStream(foundImageStream)
+		err = client.Update(context.TODO(), foundImageStream)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Build a Registry ImageStream as desired for the source cluster.
+func (t *Task) buildRegistryImageStream(name string) (*imagev1API.ImageStream, error) {
+	labels := t.PlanResources.MigPlan.GetCorrelationLabels()
+	labels["app"] = name
+	imagestream := &imagev1API.ImageStream{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:    labels,
+			Name:      name,
+			Namespace: VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryImageStream(imagestream)
+	return imagestream, err
+}
+
+// Update a Registry ImageStream as desired for the specified cluster.
+func (t *Task) updateRegistryImageStream(imagestream *imagev1API.ImageStream) error {
+	imagestream.Spec = imagev1API.ImageStreamSpec{
+		LookupPolicy: imagev1API.ImageLookupPolicy{Local: false},
+		Tags: []imagev1API.TagReference{
+			imagev1API.TagReference{
+				Name: "2",
+				Annotations: map[string]string{
+					"openshift.io/imported-from": "registry:2",
```

Comment:
Looks like we are hard-coding the "imported-from" registry name.  In what ways, if any, does this limit us?

Reply:
I'm not sure. I"m following the example of `oc new-app` functionality. I'm open to removing it, but I was trying to stick with what I know already works.

Reply:
So the `registry:2` part is from oc new-app?